### PR TITLE
udp: add explicit config for GRO

### PR DESCRIPTION
**Pull request by mattklein123**
GRO is now configurable and defaults to disabled for server sockets and
enabled for client sockets.

Fixes https://github.com/envoyproxy/envoy/issues/15418

Risk Level: Low
Testing: Updated tests
Docs Changes: None
Release Notes: Added
Platform Specific Features: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
cc @ggreenway @alyssawilk

File: source/common/network/utility.cc [hunk 1]
```diff
@@ -724,5 +726,11 @@ Api::IoErrorPtr Utility::readPacketsFromSocket(IoHandle& handle,
   } while (true);
 }
 
+ResolvedUdpSocketConfig::ResolvedUdpSocketConfig(
```
Comment (alyssawilk) [inline source/common/network/utility.cc:729]: do we log a warning if GRO is configured on and the OS doesn't support it?

File: api/envoy/config/core/v3/udp_socket_config.proto [hunk 2]
```diff
@@ -21,4 +21,10 @@ message UdpSocketConfig {
   // defaults to 1500 bytes.
   google.protobuf.UInt64Value max_rx_datagram_size = 1
       [(validate.rules).uint64 = {lt: 65536 gt: 0}];
+
+  // Configures whether `Generic Receive Offload (GRO)
+  // <https://en.wikipedia.org/wiki/Large_receive_offload>`_ is enabled (if the operating system
+  // supports it) when reading from the UDP socket. The default is context dependent and is
+  // documented where UdpSocketConfig is used.
+  google.protobuf.BoolValue use_gro = 2;
```
Comment (alyssawilk) [inline api/envoy/config/core/v3/udp_socket_config.proto:29]: I thought GRO was still compiled as an extension. 
Do we need to do the compile options here or do checks somewhere that the GRO code is included as part of config validation?

**Review by alyssawilk — COMMENTED** [review]
Looks good.  Please check if coverage failure is real as well as the nits below!

File: api/envoy/config/core/v3/udp_socket_config.proto (hunk 2 above)
Comment (mattklein123) [reply api/envoy/config/core/v3/udp_socket_config.proto:29]: No GRO is always in, GSO is the extension right now which we (you :)) need to fix.

Comment (alyssawilk) [reply api/envoy/config/core/v3/udp_socket_config.proto:29]: Right, of course just testing you, I would never get my UDP acronyms confused ;-)
(haha, thanks)

**Review by alyssawilk — APPROVED** [review]
LGTM modulo greenway and CI being happy :-)

**Review by ggreenway — CHANGES_REQUESTED** [review]
Need to document that if GRO is not available/supported, it will fallback to non-GRO, not fail to load config. Consider renaming to `prefer_gro` or `use_gro_if_supported`, but at least document this behavior clearly.

**Comment by ggreenway** [comment]
> Need to document that if GRO is not available/supported, it will fallback to non-GRO, not fail to load config. Consider renaming to `prefer_gro` or `use_gro_if_supported`, but at least document this behavior clearly.

Maybe also clarify that this config does not affect correctness, just performance, which is why it's fine to fallback if what the user specified isn't available.

**Comment by ggreenway** [comment]
Maybe this: 

> Configures whether `Generic Receive Offload (GRO)  <https://en.wikipedia.org/wiki/Large_receive_offload>`_ is preferred when reading from the UDP socket. The default is context dependent and is documented where UdpSocketConfig is used. This option affects performance but not functionality. If GRO is not supported by the operating system, non-GRO receive will be used.

**Comment by mattklein123** [comment]
@ggreenway updated per comments.

**Review by ggreenway — APPROVED** [review]
LGTM

**Pull request merged**